File: lua/puzzle1-2.lua
local sum = 0
local digits = {one=1, two=2, three=3, four=4, five=5, six=6, seven=7, eight=8, nine=9}
for line in io.lines() do
	local p1 = string.find(line, "%d")
	local d1
	if p1 then d1 = string.sub(line, p1, p1) end
	for k, v in pairs(digits) do
		local p = string.find(line, k)
		if p then
			if not p1 or p < p1 then
				p1 = p
				d1 = v
			end
		end
	end
	assert(d1)

	local d2
	for i = #line, 1, -1 do
		if string.find(line, "%d", i) then
			d2 = string.sub(line, i, i)
			break
		end
		for k, v in pairs(digits) do
			local p = string.find(line, k, i)
			if p then
				d2 = v
				break
			end
		end
		if d2 then break end
	end
	assert(d2)

	sum = sum + (d1 .. d2)
end
print(sum)

File: rust/src/bin/puzzle1-2.rs
fn main() {
    let mut sum: u32 = 0;
    let digits = &[
        "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
    ];
    for line in std::io::stdin().lines().map(|r| r.unwrap()) {
        let mut min = line.find(|c: char| c.is_ascii_digit()).unwrap();
        let mut digit1 = char::from(line.as_bytes()[min]).to_digit(10).unwrap();
        for (i, digit) in digits.iter().enumerate() {
            if let Some(pos) = line.find(digit) {
                if pos < min {
                    min = pos;
                    digit1 = (i + 1) as u32;
                }
            }
        }

        let mut max = line.rfind(|c: char| c.is_ascii_digit()).unwrap();
        let mut digit2 = char::from(line.as_bytes()[max]).to_digit(10).unwrap();
        for (i, digit) in digits.iter().enumerate() {
            if let Some(pos) = line.rfind(digit) {
                if pos > max {
                    max = pos;
                    digit2 = (i + 1) as u32;
                }
            }
        }
        sum += digit1 * 10 + digit2;
    }
    println!("{sum}");
}
